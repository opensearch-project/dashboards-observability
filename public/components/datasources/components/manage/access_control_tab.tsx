/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiText,
  EuiHorizontalRule,
} from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { EuiPanel } from '@elastic/eui';
import { ConnectionManagementCallout } from './connection_management_callout';
import { coreRefs } from '../../../../framework/core_refs';
import { QueryPermissionsConfiguration } from './query_permissions';
import { DATACONNECTIONS_BASE } from '../../../../../common/constants/shared';
import { SaveOrCancel } from '../save_or_cancel';
import { Role } from '../../../../../common/types/data_connections';

interface AccessControlTabProps {
  dataConnection: string;
  connector: string;
  properties: unknown;
  allowedRoles: string[];
}

export const AccessControlTab = (props: AccessControlTabProps) => {
  const [mode, setMode] = useState<'view' | 'edit'>('view');
  const [roles, setRoles] = useState<Role[]>([]);
  const [selectedQueryPermissionRoles, setSelectedQueryPermissionRoles] = useState<Role[]>(
    props.allowedRoles.map((role) => {
      return { label: role };
    })
  );
  const { http } = coreRefs;

  useEffect(() => {
    http!.get('/api/v1/configuration/roles').then((data) =>
      setRoles(
        Object.keys(data.data).map((key) => {
          return { label: key };
        })
      )
    );
  }, []);

  const AccessControlDetails = () => {
    return (
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Query access</EuiText>
              <EuiText size="s" className="overview-content">
                {selectedQueryPermissionRoles.length ? `Restricted` : '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const EditAccessControlDetails = () => {
    return (
      <EuiFlexGroup direction="column">
        <QueryPermissionsConfiguration
          roles={roles}
          selectedRoles={selectedQueryPermissionRoles}
          setSelectedRoles={setSelectedQueryPermissionRoles}
          layout={'vertical'}
        />
      </EuiFlexGroup>
    );
  };

  const saveChanges = () => {
    http!.put(`${DATACONNECTIONS_BASE}`, {
      body: JSON.stringify({
        name: props.dataConnection,
        allowedRoles: selectedQueryPermissionRoles.map((role) => role.label),
        connector: props.connector,
        properties: props.properties,
      }),
    });
    setMode('view');
  };

  const AccessControlHeader = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText size="m">
            <h2 className="panel-title">Access Control</h2>
            Control which OpenSearch users have access to this data source.
          </EuiText>
        </EuiFlexItem>

        <EuiFlexItem grow={false}>
          <EuiButton
            data-test-subj="createButton"
            onClick={() => setMode(mode === 'view' ? 'edit' : 'view')}
            fill={mode === 'view' ? true : false}
          >
            {mode === 'view' ? 'Edit' : 'Cancel'}
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  return (
    <>
      <EuiSpacer />
      <ConnectionManagementCallout />
      <EuiSpacer />
      <EuiPanel>
        <AccessControlHeader />
        <EuiHorizontalRule />
        {mode === 'view' ? <AccessControlDetails /> : <EditAccessControlDetails />}
      </EuiPanel>
      <EuiSpacer />
      {mode === 'edit' && (
        <SaveOrCancel
          onCancel={() => {
            setMode('view');
          }}
          onSave={saveChanges}
        />
      )}
      <EuiSpacer />
    </>
  );
};
