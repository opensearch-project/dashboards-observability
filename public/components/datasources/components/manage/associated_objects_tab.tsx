/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from 'react';
import {
  EuiInMemoryTable,
  EuiLink,
  EuiPanel,
  EuiFlexGroup,
  EuiFlexItem,
  EuiText,
  EuiHorizontalRule,
  EuiButton,
  EuiSpacer,
  EuiEmptyPrompt,
} from '@elastic/eui';
import { AssociatedObject } from 'common/types/data_connections';
import { AccelerationsRecommendationCallout } from './accelerations_recommendation_callout';

interface AssociatedObjectsTabProps {
  associatedObjects: AssociatedObject[];
}

export const AssociatedObjectsTab: React.FC<AssociatedObjectsTabProps> = ({
  associatedObjects,
}) => {
  const [lastUpdated, setLastUpdated] = useState('');

  // TODO: FINISH THE REFRESH LOGIC
  const fetchAssociatedObjects = async () => {
    // Placeholder for data fetching logic
    // After fetching data:
    // setAssociatedObjects(fetchedData);
    const now = new Date();
    setLastUpdated(now.toUTCString()); // Update last updated time
  };

  useEffect(() => {
    fetchAssociatedObjects();
  }, []);

  const AssociatedObjectsHeader = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText size="m">
            <h2 className="panel-title">Associated objects</h2>
            Manage objects associated with this data sources.
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <div style={{ textAlign: 'right' }}>
            <EuiText color="subdued" style={{ fontSize: 'small' }}>
              Last updated at:
            </EuiText>
            <EuiText color="subdued" style={{ fontSize: 'small' }}>
              {lastUpdated}
            </EuiText>
          </div>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            data-test-subj="freshButton"
            iconType="refresh"
            onClick={fetchAssociatedObjects}
          >
            Refresh
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const noDataMessage = (
    <EuiEmptyPrompt
      title={<h2>You have no associated objects</h2>}
      body={<p>Add or config tables from your data source or use Query Workbench.</p>}
      actions={
        <EuiButton
          color="primary"
          fill
          onClick={() => window.open('https://example.com', '_blank')}
          iconType="popout"
          iconSide="left"
        >
          Query Workbench
        </EuiButton>
      }
    />
  );

  const columns = [
    {
      field: 'name',
      name: 'Name',
      sortable: true,
      'data-test-subj': 'nameCell',
      render: (name: string) => (
        <EuiLink href="https://example.com" target="_blank">
          {name}
        </EuiLink>
      ),
    },
    {
      field: 'database',
      name: 'Database',
      truncateText: true,
      render: (database: string) => (
        <EuiLink href="https://example.com" target="_blank">
          {database}
        </EuiLink>
      ),
    },
    {
      field: 'type',
      name: 'Type',
      sortable: true,
    },
    {
      field: 'createdByIntegration',
      name: 'Created by Integration',
      sortable: true,
    },
    {
      field: 'accelerations',
      name: 'Accelerations',
      sortable: true,
    },
    {
      name: 'Actions',
      actions: [
        {
          name: 'Edit',
          description: 'Edit this object',
          type: 'icon',
          icon: 'discoverApp',
          onClick: (item: AssociatedObject) => console.log('Edit', item),
        },
        {
          name: 'Delete',
          description: 'Delete this object',
          type: 'icon',
          icon: 'bolt',
          onClick: (item: AssociatedObject) => console.log('Delete', item),
        },
      ],
    },
  ];

  const search = {
    box: {
      incremental: true,
      schema: {
        fields: { name: { type: 'string' }, database: { type: 'string' } },
      },
    },
  };

  const pagination = {
    initialPageSize: 10,
    pageSizeOptions: [10, 25, 50],
  };

  const sorting = {
    sort: {
      field: 'name',
      direction: 'asc',
    },
  };

  return (
    <>
      <EuiSpacer />
      <EuiPanel>
        <AssociatedObjectsHeader />
        <EuiHorizontalRule />
        <AccelerationsRecommendationCallout />
        <EuiSpacer />
        {associatedObjects.length > 0 ? (
          <EuiInMemoryTable
            items={associatedObjects}
            columns={columns}
            search={search}
            pagination={pagination}
            sorting={sorting}
            noItemsMessage={associatedObjects.length === 0 ? noDataMessage : undefined}
          />
        ) : (
          noDataMessage
        )}
      </EuiPanel>
      <EuiSpacer />
    </>
  );
};
