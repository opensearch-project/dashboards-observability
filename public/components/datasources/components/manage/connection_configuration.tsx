/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiFieldPassword,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSelect,
  EuiSpacer,
  EuiText,
  EuiTextArea,
} from '@elastic/eui';
import { EuiSelectOption } from '@elastic/eui/src/components/form/select';
import React, { useState } from 'react';

interface ConnectionConfigurationProps {
  connectionName: string;
  connectionDetails: string;
  onConnectionDetailsChange: (e: any) => void;
  authenticationOptions: EuiSelectOption[];
  setSelectedAuthenticationMethod: (authenticationMethod: EuiSelectOption) => void;
  selectedAuthenticationMethod: string;
}

export const ConnectionConfiguration = (props: ConnectionConfigurationProps) => {
  const {
    connectionName,
    connectionDetails,
    onConnectionDetailsChange,
    authenticationOptions,
    selectedAuthenticationMethod,
    setSelectedAuthenticationMethod,
  } = props;
  const [details, setDetails] = useState(connectionDetails);

  const [password, setPassword] = useState('');

  const NameRow = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Data source name</EuiText>
          <EuiText size="s" className="overview-content">
            This is the name of the data source and how it will be referenced in OpenSearch
            Dashboards.
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Data source name">
            <EuiFieldText placeholder={connectionName} readOnly />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const SparkEndpointRow = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Spark endpoint URL</EuiText>
          <EuiText size="s" className="overview-content">
            {
              "The URL for your Spark cluster and where your data is. This is what OpenSearch will connect to. The endpoint URL can't be changed. If you'd like to use another endpoint create a new data source."
            }
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Spark endpoint URL">
            <EuiFieldText placeholder={'-'} readOnly />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  return (
    <EuiFlexItem>
      <NameRow />
      <EuiSpacer />
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Description - optional</EuiText>
          <EuiText size="s" className="overview-content">
            Text that can help identify the data source or share additional details
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Description">
            <EuiTextArea
              placeholder="Placeholder text"
              aria-label="test"
              value={details}
              onChange={(e) => {
                setDetails(e.target.value);
              }}
              onBlur={onConnectionDetailsChange}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer />
      <SparkEndpointRow />
      <EuiSpacer />
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Authentication details</EuiText>
          <EuiText size="s" className="overview-content">
            This is information used to authenticate and create a data source with Spark.
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Authentication method">
            <EuiSelect
              options={authenticationOptions}
              value={selectedAuthenticationMethod}
              onChange={(e) => setSelectedAuthenticationMethod(e)}
            />
          </EuiFormRow>
          <EuiFormRow label="Username">
            <EuiFieldText placeholder={'Username placeholder'} />
          </EuiFormRow>
          <EuiFormRow label="Password">
            <EuiFieldPassword
              type="dual"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiFlexItem>
  );
};
