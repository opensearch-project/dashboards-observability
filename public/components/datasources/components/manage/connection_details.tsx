/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiFlexGroup, EuiFlexItem, EuiSpacer, EuiText, EuiHorizontalRule } from '@elastic/eui';
import React, { useState } from 'react';
import { EuiPanel } from '@elastic/eui';
import { ConnectionManagementCallout } from './connection_management_callout';
import { PrometheusProperties, S3GlueProperties } from './data_connection';
import { DatasourceType } from '../../../../../common/types/data_connections';

interface ConnectionDetailProps {
  dataConnection: string;
  connector: DatasourceType;
  description: string;
  properties: S3GlueProperties | PrometheusProperties;
}

export const ConnectionDetails = (props: ConnectionDetailProps) => {
  const { dataConnection, connector, description, properties } = props;

  const S3ConnectionConfigurationView = () => {
    return (
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Data source name</EuiText>
              <EuiText size="s" className="overview-content">
                {dataConnection}
              </EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Data source description</EuiText>
              <EuiText size="s" className="overview-content">
                {description || '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Index store region</EuiText>
              <EuiText size="s" className="overview-content">
                {(properties as S3GlueProperties)['glue.indexstore.opensearch.region'] || '-'}
              </EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Index store URI</EuiText>
              <EuiText size="s" className="overview-content">
                {(properties as S3GlueProperties)['glue.indexstore.opensearch.uri'] || '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const PrometheusConnectionConfigurationView = () => {
    return (
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Data source name</EuiText>
              <EuiText size="s" className="overview-content">
                {dataConnection}
              </EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Data source description</EuiText>
              <EuiText size="s" className="overview-content">
                {description || '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Prometheus URI</EuiText>
              <EuiText size="s" className="overview-content">
                {(properties as PrometheusProperties)['prometheus.uri'] || '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const ConnectionConfigurationHeader = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText size="m">
            <h2 className="panel-title">Data source configurations</h2>
            Control configurations for your data source.
          </EuiText>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  return (
    <>
      <EuiSpacer />
      <ConnectionManagementCallout />
      <EuiSpacer />
      <EuiPanel>
        <ConnectionConfigurationHeader />
        <EuiHorizontalRule />
        {connector === 'S3GLUE' ? (
          <S3ConnectionConfigurationView />
        ) : (
          <PrometheusConnectionConfigurationView />
        )}
      </EuiPanel>
      <EuiSpacer />
      <EuiSpacer />
    </>
  );
};
