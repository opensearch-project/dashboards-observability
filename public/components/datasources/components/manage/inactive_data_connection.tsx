/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton, EuiCallOut } from '@elastic/eui';
import React from 'react';
import {
  DATACONNECTIONS_BASE,
  DATACONNECTIONS_UPDATE_STATUS,
  EDIT,
} from '../../../../../common/constants/shared';
import { coreRefs } from '../../../../framework/core_refs';
import { useToast } from '../../../common/toast';
import { DatasourceDetails } from './data_connection';

interface InactiveDataConnectionCalloutProps {
  datasourceDetails: DatasourceDetails;
  fetchSelectedDatasource: () => void;
}

export const InactiveDataConnectionCallout = ({
  datasourceDetails,
  fetchSelectedDatasource,
}: InactiveDataConnectionCalloutProps) => {
  const { setToast } = useToast();
  const { http } = coreRefs;
  const enableDataSource = () => {
    http!
      .post(`${DATACONNECTIONS_BASE}${EDIT}${DATACONNECTIONS_UPDATE_STATUS}`, {
        body: JSON.stringify({ name: datasourceDetails.name, status: 'active' }),
      })
      .then(() => {
        setToast(`Data connection ${datasourceDetails.name} enabled successfully`);
        fetchSelectedDatasource();
      })
      .catch((err) => {
        console.error(err);
        setToast(`Data connection ${datasourceDetails.name} could not be enabled.`, 'danger');
      });
  };

  return (
    <EuiCallOut title="This data source connection is inactive" color="warning" iconType="help">
      <p>
        Associated objects and accelerations are not available while this connection is inactive.
      </p>
      <EuiButton onClick={enableDataSource} color="warning">
        Enable connection
      </EuiButton>
    </EuiCallOut>
  );
};
