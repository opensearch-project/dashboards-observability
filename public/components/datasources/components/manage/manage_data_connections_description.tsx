/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiSpacer,
  EuiText,
  EuiTitle,
  EuiHorizontalRule,
  EuiLink,
  EuiSmallButton,
  EuiFlexGroup,
  EuiFlexItem,
} from '@elastic/eui';
import React from 'react';

interface DataConnectionsDescriptionProps {
  refresh: () => void;
}

export const DataConnectionsDescription = (props: DataConnectionsDescriptionProps) => {
  const { refresh } = props;
  return (
    <div>
      <EuiFlexGroup justifyContent="spaceBetween">
        <EuiFlexItem grow={false}>
          <>
            <EuiTitle size="s">
              <h2>Manage data sources</h2>
            </EuiTitle>

            <EuiSpacer size="s" />
            <EuiText size="s" color="subdued">
              Manage existing data sources or{' '}
              <EuiLink href="#/new">create a new data source</EuiLink>
            </EuiText>
          </>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiSmallButton onClick={refresh} iconType="refresh">
            Refresh
          </EuiSmallButton>
        </EuiFlexItem>
      </EuiFlexGroup>

      <EuiHorizontalRule size="full" />
    </div>
  );
};
