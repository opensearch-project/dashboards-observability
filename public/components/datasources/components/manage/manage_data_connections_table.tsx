/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiHealth,
  EuiIcon,
  EuiInMemoryTable,
  EuiLink,
  EuiOverlayMask,
  EuiPage,
  EuiPageBody,
  EuiPageContent,
  EuiTableFieldDataColumnType,
} from '@elastic/eui';
import _ from 'lodash';
import React, { useEffect, useState } from 'react';
import {
  DATACONNECTIONS_BASE,
  observabilityMetricsID,
} from '../../../../../common/constants/shared';
import {
  DatasourceDetails,
  DatasourceStatus,
  DatasourceType,
} from '../../../../../common/types/data_connections';
import { coreRefs } from '../../../../../public/framework/core_refs';
import { getRenderCreateAccelerationFlyout } from '../../../../../public/plugin';
import { DeleteModal } from '../../../common/helpers/delete_modal';
import { useToast } from '../../../common/toast';
import { HomeProps } from '../../home';
import PrometheusLogo from '../../icons/prometheus-logo.svg';
import S3Logo from '../../icons/s3-logo.svg';
import { DataConnectionsHeader } from '../data_connections_header';
import { redirectToExplorerS3 } from './associated_objects/utils/associated_objects_tab_utils';
import { InstallIntegrationFlyout } from './integrations/installed_integrations_table';
import { DataConnectionsDescription } from './manage_data_connections_description';

interface DataConnection {
  connectionType: DatasourceType;
  name: string;
  dsStatus: DatasourceStatus;
}

export const ManageDataConnectionsTable = (props: HomeProps) => {
  const { http, chrome } = props;
  const { application } = coreRefs;

  const { setToast } = useToast();

  const [data, setData] = useState<DataConnection[]>([]);
  const [isModalVisible, setIsModalVisible] = useState(false);
  const [modalLayout, setModalLayout] = useState(<EuiOverlayMask />);

  const deleteConnection = (connectionName: string) => {
    http!
      .delete(`${DATACONNECTIONS_BASE}/${connectionName}`)
      .then(() => {
        setToast(`Data connection ${connectionName} deleted successfully`);
        setData(
          data.filter((connection) => {
            return !(connection.name === connectionName);
          })
        );
      })
      .catch((err) => {
        console.error(err);
        setToast(`Data connection $${connectionName} not deleted. See output for more details.`);
      });
  };

  const fetchDataSources = () => {
    http!
      .get(`${DATACONNECTIONS_BASE}`)
      .then((res: DatasourceDetails[]) => {
        const dataConnections = res.map((dataConnection: DatasourceDetails) => {
          return {
            name: dataConnection.name,
            connectionType: dataConnection.connector,
            dsStatus: dataConnection.status,
          };
        });
        setData(dataConnections);
      })
      .catch((err) => {
        console.error(err);
        setToast(`Could not fetch datasources`, 'danger');
      });
  };

  useEffect(() => {
    chrome.setBreadcrumbs([
      {
        text: 'Data sources',
        href: '#/',
      },
    ]);
    fetchDataSources();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chrome]);

  const displayDeleteModal = (connectionName: string) => {
    setModalLayout(
      <DeleteModal
        onConfirm={() => {
          setIsModalVisible(false);
          deleteConnection(connectionName);
        }}
        onCancel={() => {
          setIsModalVisible(false);
        }}
        title={`Delete ${connectionName}`}
        message={`Are you sure you want to delete ${connectionName}?`}
      />
    );
    setIsModalVisible(true);
  };

  const [showIntegrationsFlyout, setShowIntegrationsFlyout] = useState(false);
  const [integrationsFlyout, setIntegrationsFlyout] = useState<React.JSX.Element | null>(null);

  const actions = [
    {
      name: (datasource: DataConnection) =>
        `Query in ${
          datasource.connectionType === 'PROMETHEUS' ? 'Metrics Analytics' : 'Observability Logs'
        }`,
      isPrimary: true,
      icon: 'discoverApp',
      type: 'icon',
      onClick: (datasource: DataConnection) => {
        if (datasource.connectionType === 'PROMETHEUS') {
          application!.navigateToApp(observabilityMetricsID);
        } else if (datasource.connectionType === 'S3GLUE') {
          redirectToExplorerS3(datasource.name);
        }
      },
      'data-test-subj': 'action-query',
    },
    {
      name: 'Accelerate performance',
      isPrimary: false,
      icon: 'bolt',
      type: 'icon',
      available: (datasource: DataConnection) => datasource.connectionType !== 'PROMETHEUS',
      onClick: (datasource: DataConnection) => {
        renderCreateAccelerationFlyout({ dataSource: datasource.name });
      },
      'data-test-subj': 'action-accelerate',
    },
    {
      name: 'Integrate data',
      isPrimary: false,
      icon: 'integrationGeneral',
      type: 'icon',
      available: (datasource: DataConnection) => datasource.connectionType !== 'PROMETHEUS',
      onClick: (datasource: DataConnection) => {
        setIntegrationsFlyout(
          <InstallIntegrationFlyout
            closeFlyout={() => setShowIntegrationsFlyout(false)}
            datasourceType={datasource.connectionType}
            datasourceName={datasource.name}
          />
        );
        setShowIntegrationsFlyout(true);
      },
      'data-test-subj': 'action-integrate',
    },
    {
      name: 'Delete',
      description: 'Delete this data source',
      icon: 'trash',
      color: 'danger',
      type: 'icon',
      onClick: (datasource: DataConnection) => displayDeleteModal(datasource.name),
      isPrimary: false,
      'data-test-subj': 'action-delete',
    },
  ];

  const icon = (record: DataConnection) => {
    switch (record.connectionType) {
      case 'S3GLUE':
        return <EuiIcon type={S3Logo} />;
      case 'PROMETHEUS':
        return <EuiIcon type={PrometheusLogo} />;
      default:
        return <></>;
    }
  };

  const tableColumns = [
    {
      field: 'name',
      name: 'Name',
      sortable: true,
      truncateText: true,
      render: (value, record: DataConnection) => (
        <EuiFlexGroup>
          <EuiFlexItem grow={false}>{icon(record)}</EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiLink
              data-test-subj={`${record.name}DataConnectionsLink`}
              href={`#/manage/${record.name}`}
            >
              {_.truncate(record.name, { length: 100 })}
            </EuiLink>
          </EuiFlexItem>
        </EuiFlexGroup>
      ),
    },
    {
      field: 'status',
      name: 'Status',
      sortable: true,
      truncateText: true,
      render: (value, record: DataConnection) =>
        record.dsStatus === 'ACTIVE' ? (
          <EuiHealth color="success">Active</EuiHealth>
        ) : (
          <EuiHealth color="subdued">Inactive</EuiHealth>
        ),
    },
    {
      field: 'actions',
      name: 'Actions',
      actions,
    },
  ] as Array<EuiTableFieldDataColumnType<unknown>>;

  const search = {
    box: {
      incremental: true,
    },
  };

  const entries = data.map((dataconnection: DataConnection) => {
    const name = dataconnection.name;
    const connectionType = dataconnection.connectionType;
    const dsStatus = dataconnection.dsStatus;
    return { connectionType, name, dsStatus, data: { name, connectionType } };
  });

  const renderCreateAccelerationFlyout = getRenderCreateAccelerationFlyout();

  return (
    <EuiPage>
      <EuiPageBody component="div">
        <DataConnectionsHeader />
        <EuiPageContent data-test-subj="manageDataConnectionsarea">
          <DataConnectionsDescription refresh={fetchDataSources} />
          <EuiInMemoryTable
            items={entries}
            itemId="id"
            columns={tableColumns}
            tableLayout="auto"
            pagination={{
              initialPageSize: 10,
              pageSizeOptions: [5, 10, 15],
            }}
            search={search}
            allowNeutralSort={false}
            isSelectable={true}
          />
        </EuiPageContent>
        {isModalVisible && modalLayout}
        {showIntegrationsFlyout && integrationsFlyout}
      </EuiPageBody>
    </EuiPage>
  );
};
