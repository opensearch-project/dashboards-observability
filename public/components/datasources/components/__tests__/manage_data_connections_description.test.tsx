/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { waitFor } from '@testing-library/react';
import React from 'react';
import { DataConnectionsDescription } from '../manage/manage_data_connections_description';

describe('Manage Data Connections Description test', () => {
  configure({ adapter: new Adapter() });

  it('Renders manage data connections description', async () => {
    const wrapper = mount(<DataConnectionsDescription refresh={() => {}} />);

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
