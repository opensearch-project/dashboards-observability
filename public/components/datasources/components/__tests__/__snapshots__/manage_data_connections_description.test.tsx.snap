// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Data Connections Description test Renders manage data connections description 1`] = `
<Component
  refresh={[Function]}
>
  <div>
    <EuiFlexGroup
      justifyContent="spaceBetween"
    >
      <div
        className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <EuiFlexItem
          grow={false}
        >
          <div
            className="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <EuiTitle
              size="s"
            >
              <h2
                className="euiTitle euiTitle--small"
              >
                Manage data sources
              </h2>
            </EuiTitle>
            <EuiSpacer
              size="s"
            >
              <div
                className="euiSpacer euiSpacer--s"
              />
            </EuiSpacer>
            <EuiText
              color="subdued"
              size="s"
            >
              <div
                className="euiText euiText--small"
              >
                <EuiTextColor
                  color="subdued"
                  component="div"
                >
                  <div
                    className="euiTextColor euiTextColor--subdued"
                  >
                    Manage existing data sources or
                     
                    <EuiLink
                      href="#/new"
                    >
                      <a
                        className="euiLink euiLink--primary"
                        href="#/new"
                        rel="noreferrer"
                      >
                        create a new data source
                      </a>
                    </EuiLink>
                  </div>
                </EuiTextColor>
              </div>
            </EuiText>
          </div>
        </EuiFlexItem>
        <EuiFlexItem
          grow={false}
        >
          <div
            className="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <EuiButton
              iconType="refresh"
              onClick={[Function]}
            >
              <EuiButtonDisplay
                baseClassName="euiButton"
                disabled={false}
                element="button"
                iconType="refresh"
                isDisabled={false}
                onClick={[Function]}
                type="button"
              >
                <button
                  className="euiButton euiButton--primary"
                  disabled={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "minWidth": undefined,
                    }
                  }
                  type="button"
                >
                  <EuiButtonContent
                    className="euiButton__content"
                    iconSide="left"
                    iconType="refresh"
                    textProps={
                      Object {
                        "className": "euiButton__text",
                      }
                    }
                  >
                    <span
                      className="euiButtonContent euiButton__content"
                    >
                      <EuiIcon
                        className="euiButtonContent__icon"
                        color="inherit"
                        size="m"
                        type="refresh"
                      >
                        <EuiIconEmpty
                          aria-hidden={true}
                          className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                          focusable="false"
                          role="img"
                          style={null}
                        >
                          <svg
                            aria-hidden={true}
                            className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                            focusable="false"
                            height={16}
                            role="img"
                            style={null}
                            viewBox="0 0 16 16"
                            width={16}
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </EuiIconEmpty>
                      </EuiIcon>
                      <span
                        className="euiButton__text"
                      >
                        Refresh
                      </span>
                    </span>
                  </EuiButtonContent>
                </button>
              </EuiButtonDisplay>
            </EuiButton>
          </div>
        </EuiFlexItem>
      </div>
    </EuiFlexGroup>
    <EuiHorizontalRule
      size="full"
    >
      <hr
        className="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginLarge"
      />
    </EuiHorizontalRule>
  </div>
</Component>
`;
